### PR TITLE
Const char handlers

### PR DESCRIPTION
**Pull request by erik-johnson-nimbelink**
Updated some C string uses to be const char instead of just char to make passing in const char a little bit easier -- especially since the APIs shouldn't be touching the buffer contents anyway. Might as well contractually say that in the signature.

### CODE REVIEW

**Comment by sigvartmh** [comment]
nit: You could newline parameters in  dfu_ctx_mcuboot_set_b1_file to please checkpatch,

**Comment by rlubos** [comment]
Hi @erik-johnson-nimbelink,
It seems that several functions still miss the `const` qualifier, I've summarized the CI failures below, I hope that I have not missed any of them.

```
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c: In function 'slm_at_host_init':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:372:40: error: passing argument 2 of 'at_notif_register_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
  err = at_notif_register_handler(NULL, response_handler);
                                        ^~~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:15:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/include/at_notif.h:64:65: note: expected 'at_notif_handler_t' {aka 'void (*)(void *, const char *)'} but argument is of type 'void (*)(void *, char *)'
 int at_notif_register_handler(void *context, at_notif_handler_t handler);
                                              ~~~~~~~~~~~~~~~~~~~^~~~~~~
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:392:26: error: passing argument 1 of 'slm_at_tcpip_init' from incompatible pointer type [-Werror=incompatible-pointer-types]
  err = slm_at_tcpip_init(slm_at_callback);
                          ^~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:20:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_tcpip.h:37:40: note: expected 'at_cmd_handler_t' {aka 'void (*)(const char *)'} but argument is of type 'void (*)(char *)'
 int slm_at_tcpip_init(at_cmd_handler_t callback);
                       ~~~~~~~~~~~~~~~~~^~~~~~~~
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:400:24: error: passing argument 1 of 'slm_at_gps_init' from incompatible pointer type [-Werror=incompatible-pointer-types]
  err = slm_at_gps_init(slm_at_callback);
                        ^~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:24:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_gps.h:37:38: note: expected 'at_cmd_handler_t' {aka 'void (*)(const char *)'} but argument is of type 'void (*)(char *)'
 int slm_at_gps_init(at_cmd_handler_t callback);
                     ~~~~~~~~~~~~~~~~~^~~~~~~~
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c: In function 'slm_at_host_uninit':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:428:42: error: passing argument 2 of 'at_notif_deregister_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
  err = at_notif_deregister_handler(NULL, response_handler);
                                          ^~~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:15:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/include/at_notif.h:79:67: note: expected 'at_notif_handler_t' {aka 'void (*)(void *, const char *)'} but argument is of type 'void (*)(void *, char *)'
 int at_notif_deregister_handler(void *context, at_notif_handler_t handler);
```

```
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/bin/lwm2m_carrier/os/lwm2m_os.c: In function 'lwm2m_os_at_notif_register_handler':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/bin/lwm2m_carrier/os/lwm2m_os.c:343:44: error: passing argument 2 of 'at_notif_register_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
  return at_notif_register_handler(context, handler);
                                            ^~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/bin/lwm2m_carrier/os/lwm2m_os.c:14:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/include/at_notif.h:64:65: note: expected 'at_notif_handler_t' {aka 'void (*)(void *, const char *)'} but argument is of type 'lwm2m_os_at_cmd_handler_t' {aka 'void (*)(void *, char *)'}
 int at_notif_register_handler(void *context, at_notif_handler_t handler);
```

```
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/at_host/at_host.c: In function 'at_host_init':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/at_host/at_host.c:286:40: error: passing argument 2 of 'at_notif_register_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
  err = at_notif_register_handler(NULL, response_handler);
                                        ^~~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/at_host/at_host.c:15:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/include/at_notif.h:64:65: note: expected 'at_notif_handler_t' {aka 'void (*)(void *, const char *)'} but argument is of type 'void (*)(void *, char *)'
 int at_notif_register_handler(void *context, at_notif_handler_t handler);
```

```
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/modem_info/modem_info.c: In function 'modem_info_rsrp_register':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/modem_info/modem_info.c:530:3: error: passing argument 2 of 'at_notif_register_handler' from incompatible pointer type [-Werror=incompatible-pointer-types]
   modem_info_rsrp_subscribe_handler);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/lib/modem_info/modem_info.c:9:
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master@2/sources/nrf/include/at_notif.h:64:65: note: expected 'at_notif_handler_t' {aka 'void (*)(void *, const char *)'} but argument is of type 'void (*)(void *, char *)'
 int at_notif_register_handler(void *context, at_notif_handler_t handler);
```

**Comment by erik-johnson-nimbelink** [comment]
I haven't really found a way to build everything, so I figured there was a chance I would have missed some. I'll go through and correct all of those here in a bit. Thanks for letting me know about them.

Let me know if there's a way for me to handle catching any others! I'm happy to do that work myself.

File: lib/bin/lwm2m_carrier/include/lwm2m_os.h [hunk 1]
```diff
@@ -54,7 +54,7 @@ struct lwm2m_os_at_param_list {
 /**
  * @brief AT Command handler.
  */
-typedef void (*lwm2m_os_at_cmd_handler_t)(void *ctx, char *response);
```
Comment (lemrey) [inline lib/bin/lwm2m_carrier/include/lwm2m_os.h]: Not sure you can change this file without release a new `liblwm2m_carrier` library.

**Comment by rlubos** [comment]
@erik-johnson-nimbelink The CI results are only available on our internal servers, which I find very unfortunate... Nothing we can do about right now.

Anyway, one more failure showed up in the CI:
```
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master/sources/nrf/lib/at_host/at_host.c: In function 'response_handler':
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master/sources/nrf/lib/at_host/at_host.c:81:20: error: passing argument 1 of 'write_uart_string' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
  write_uart_string(response);
                    ^~~~~~~~
/jenkins_workspace/workspace/latest/sub/test-fw-nrfconnect-nrf/master/sources/nrf/lib/at_host/at_host.c:68:44: note: expected 'char *' but argument is of type 'const char *'
 static inline void write_uart_string(char *str)
```

**Comment by erik-johnson-nimbelink** [comment]
@rlubos I took care of that last error. Hopefully that's the last one!

@lemrey Yeah, I thought about that. It looks like the source is available, but it's typically linked as a library, right? (If I remember correctly, it gets certified by carriers.)

I figured I'd make the change anyway, and let you all decide how/if you wanted to manage it. At the very least, the resulting functions should be compatible as far as calls into them go (i.e. pointer sizes should be the same).

The only risk should be the library trying to change the char * buffer's contents, which obviously would break a const char * call with something like a flash-stored string. But, looking at the source, that shouldn't happen:

int lwm2m_os_at_notif_register_handler(void *context,
                       lwm2m_os_at_cmd_handler_t handler)
{
    return at_notif_register_handler(context, handler);
}

A commit in this PR also changes the signature for the at_notif_register_handler() API to be const char *, and there weren't any issues from that change, as far as how the library uses the string. So, there shouldn't be any issues with someone linking against the previous build and using a constant string.

**Comment by rlubos** [comment]
```
/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c: In function 'response_handler':

/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:90:21: error: passing argument 1 of 'write_uart_string' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]

   write_uart_string(response, len);

                     ^~~~~~~~

/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:73:44: note: expected 'char *' but argument is of type 'const char *'

 static inline void write_uart_string(char *str, size_t len)

                                      ~~~~~~^~~

/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c: In function 'slm_at_callback':

/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:339:20: error: passing argument 1 of 'write_uart_string' discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]

  write_uart_string(str, strlen(str));

                    ^~~

/jenkins_workspace/workspace/_test-fw-nrfconnect-nrf_master_8/sources/nrf/samples/nrf9160/serial_lte_modem/src/slm_at_host.c:73:44: note: expected 'char *' but argument is of type 'const char *'

 static inline void write_uart_string(char *str, size_t len)
```

I hope that's really the last one (serial_lte_modem is the last sample failing).

**Comment by erik-johnson-nimbelink** [comment]
Ah, yeah, I was worried that one would complain too, but forgot to update it.

**Comment by lemrey** [comment]
@erik-johnson-nimbelink I agree, it should be fine as long as the linker is happy.

**Pull request merged**